hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -131,17 +182,25 @@ public List<R> consume(
 
 			final ResultHandler<R> resultHandlerToUse;
 
-			if ( uniqueSemantic == UniqueSemantic.ALLOW
-					&& domainResultJavaType instanceof EntityJavaType ) {
+			boolean isEnityResultType = domainResultJavaType instanceof EntityJavaType;
+			if ( uniqueSemantic == UniqueSemantic.ALLOW && isEnityResultType ) {
 				resultHandlerToUse = ListResultsConsumer::deDuplicationHandling;
 			}
 			else {
 				resultHandlerToUse = this.resultHandler;
 			}
 
+			final Results<R> results;
+			if ( isEnityResultType ) {
```

Comment:
we want to apply the check optimisation even when duplication is not allowed

Reply:
My point is, that there is no need to use a `EntityResult` when the `uniqueSemantic` has a value, which will always result in the use of `Result#add` instead of `addUnique`. So IMO the only cases when it makes sense to use `EntityResult` is when the `uniqueSemantic` is `ALLOW` or `FILTER` (both will use `Result#addUnique`) and the result type is an entity. `ASSERT` also uses `Result#addUnique`, but is only used for `SingleIdLoadPlan` when the list is guaranteed to have only a single element. Other `uniqueSemantic` use `Result#add`, so a `EntityResult` would only mean overhead for these scenarios.

Reply:
it seems strange the use of ASSERT for  SingleIdLoadPlan

Reply:
i would have chosen NEVER

Reply:
You were the one who chose `ASSERT` originally ;)
I think I would also prefer `NEVER` and do the "assertion" in the `SingleIdLoadPlan` code instead, but not sure if we can remove `ASSERT` now (since it will be unused then) as this is in SPI.